frontend/packages/console-dynamic-plugin-sdk/src/app/k8s/api-discovery/discovery-cache.ts

```diff
@@ -0,0 +1,58 @@
+import * as _ from 'lodash';
+
+const SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY = 'sdk/api-discovery-resources';
+
+export const cacheResources = (resources) =>
+  new Promise<void>((resolve, reject) => {
+    try {
+      // Add the console version. We invalidate the cache when console version changes.
+      const { consoleVersion } = window.SERVER_FLAGS;
+      const versionedResources = _.assign({}, resources, { consoleVersion });
+      localStorage.setItem(
+        SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY,
+        JSON.stringify(versionedResources),
+      );
+      resolve();
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.error('Error caching API resources in localStorage', e);
+      reject(e);
+    }
+  });
+
+export const getCachedResources = () =>
+  new Promise<any>((resolve, reject) => {
+    try {
+      const resourcesJSON = localStorage.getItem(SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY);
+
+      // Clear cached resources after load as a safeguard. If there's any errors
+      // with the content that prevents the console from working, the bad data
+      // will not be loaded when the user refreshes the console. The cache will
+      // be refreshed when discovery completes.
+      localStorage.removeItem(SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY);
+
+      if (resourcesJSON) {
+        const resources = JSON.parse(resourcesJSON);
+        const { consoleVersion: currentVersion } = window.SERVER_FLAGS;
+        const { consoleVersion: cachedVersion } = resources;
+        if (cachedVersion !== currentVersion) {
+          // eslint-disable-next-line no-console
+          console.log(
+            `Invalidating API discovery cache from earlier console version (current: ${currentVersion}, cached: ${cachedVersion})`,
+          );
+          resolve(null);
```

Comment:
I know this was migrated from the console so not a blocker - but it seems odd to me to resolve a promise with null. I feel like this function should return the cached resources and in all other cases reject the promise with an error for clarity purposes (IE if the cached resources are out of date). 

Curious on your thoughts and not blocking at all as this will require additional logic changes.

Reply:
make sense, updated it